Common/Securities/BuyingPowerModel.cs

```diff
@@ -358,141 +358,113 @@ public virtual HasSufficientBuyingPowerForOrderResult HasSufficientBuyingPowerFo
-            var absUnitMargin = this.GetInitialMarginRequirement(parameters.Security, 1);
-            if (absUnitMargin == 0)
+            // Our max target margin allocated to this security; if targeting zero, simply return the negative of the quantity
+            var targetMarginAllocated = parameters.TargetBuyingPower * totalUseablePortfolioValue;
+            if (targetMarginAllocated == 0)
             {
-                return new GetMaximumOrderQuantityResult(0, parameters.Security.Symbol.GetZeroPriceMessage());
+                return new GetMaximumOrderQuantityResult(-parameters.Security.Holdings.Quantity, string.Empty,
+                    false);
             }
 
-            var minimumValue = absUnitMargin * parameters.Security.SymbolProperties.LotSize;
-            if (minimumValue > absFinalOrderMargin)
-            {
-                string reason = null;
-                if (!parameters.SilenceNonErrorReasons)
-                {
-                    reason = $"The target order margin {absFinalOrderMargin} is less than the minimum {minimumValue}.";
-                }
-                return new GetMaximumOrderQuantityResult(0, reason, false);
-            }
-
-            // continue iterating while we do not have enough margin for the order
-            decimal orderMargin = 0;
-            decimal orderFees = 0;
-            // compute the initial order quantity
-            var orderQuantity = absFinalOrderMargin / absUnitMargin;
+            // First pass, calculate:
+            // - Final holdings quantity
+            // - Order quantity to get to our final holdings
+            // - Total target holdings margin requirement
+            var finalHoldingsQuantity = (targetMarginAllocated / unitMargin)
+                .DiscretelyRoundBy(parameters.Security.SymbolProperties.LotSize);
+            var orderQuantity = finalHoldingsQuantity - parameters.Security.Holdings.Quantity;
+            var targetHoldingsMargin = finalHoldingsQuantity * unitMargin;
 
-            // rounding off Order Quantity to the nearest multiple of Lot Size
-            orderQuantity -= orderQuantity % parameters.Security.SymbolProperties.LotSize;
+            // Check order quantity before moving on
             if (orderQuantity == 0)
             {
-                string reason = null;
-                if (!parameters.SilenceNonErrorReasons)
+                string reason;
+                if (finalHoldingsQuantity == parameters.Security.Holdings.Quantity)
                 {
-                    reason = $"The order quantity is less than the lot size of {parameters.Security.SymbolProperties.LotSize} " +
-                             "and has been rounded to zero.";
+                    reason =
+                        $"Already at target holding {finalHoldingsQuantity} for {parameters.Security.Symbol.Value}" +
+                        $" Current holdings: {parameters.Security.Holdings.Quantity}";
```

Comment:
For performance we shouldn't create the string if we are not going to use it as before, this happens quite often I believe